### ISSUE TITLE
Allow SendGrid Sandbox mode

### ISSUE THREAD

**Issue by MvanDiemen** [thoughtbot/bamboo#329]
We are missing the sandbox mode for the sendgrid adapter. This is an extra block at the V3 request:

```
  "mail_settings": {
      "sandbox_mode": {
          "enable": true
      }
  }
```
We build this in and have a PR ready (with tests of course):
```
diff --git a/lib/bamboo/adapters/send_grid_adapter.ex b/lib/bamboo/adapters/send_grid_adapter.ex
index 49b21a8..537fcdc 100644
--- a/lib/bamboo/adapters/send_grid_adapter.ex
+++ b/lib/bamboo/adapters/send_grid_adapter.ex
@@ -33,7 +33,7 @@ defmodule Bamboo.SendGridAdapter do
 
   def deliver(email, config) do
     api_key = get_key(config)
-    body = email |> to_sendgrid_body |> Poison.encode!
+    body = email |> to_sendgrid_body(config) |> Poison.encode!
     url = [base_uri(), @send_message_path]
 
     case :hackney.post(url, headers(api_key), body, [:with_body]) do
@@ -66,6 +66,13 @@ defmodule Bamboo.SendGridAdapter do
     end
   end
 
+  defp get_sandbox_mode(config) do
+    case Map.get(config, :sandbox_mode) do
+      nil -> false
+      key -> key
+    end
+  end
+
   defp raise_api_key_error(config) do
     raise ArgumentError, """
     There was no API key set for the SendGrid adapter.
@@ -83,7 +90,7 @@ defmodule Bamboo.SendGridAdapter do
     ]
   end
 
-  defp to_sendgrid_body(%Email{} = email) do
+  defp to_sendgrid_body(%Email{} = email, config = %{}) do
     %{}
     |> put_from(email)
     |> put_personalization(email)
@@ -92,6 +99,7 @@ defmodule Bamboo.SendGridAdapter do
     |> put_content(email)
     |> put_template_id(email)
     |> put_attachments(email)
+    |> put_mail_settings(config)
   end
 
   defp put_from(body, %Email{from: from}) do
@@ -110,6 +118,10 @@ defmodule Bamboo.SendGridAdapter do
     |> put_template_substitutions(email)
   end
 
+  defp put_mail_settings(body, config) do
+    Map.put(body, :mail_settings, %{sandbox_mode: %{enable: get_sandbox_mode(config)}})
+  end
```
And the tests:
```
--- a/test/lib/bamboo/adapters/send_grid_adapter_test.exs
+++ b/test/lib/bamboo/adapters/send_grid_adapter_test.exs
@@ -4,6 +4,7 @@ defmodule Bamboo.SendGridAdapterTest do
   alias Bamboo.SendGridAdapter
 
   @config %{adapter: SendGridAdapter, api_key: "123_abc"}
+  @config_with_sandbox_mode %{adapter: SendGridAdapter, api_key: "123_abc", sandbox_mode: true}
   @config_with_bad_key %{adapter: SendGridAdapter, api_key: nil}
 
   defmodule FakeSendgrid do
@@ -77,6 +78,20 @@ defmodule Bamboo.SendGridAdapterTest do
     assert request_path == "/mail/send"
   end
 
+  test "deliver/2 defaults sandbox mode to false" do
+    new_email() |> SendGridAdapter.deliver(@config)
+
+    assert_receive {:fake_sendgrid, %{params: params}}
+    refute params["mail_settings"]["sandbox_mode"]["enable"]
+  end
+
+  test "deliver/2 allows sandbox mode enabling via config" do
+    new_email() |> SendGridAdapter.deliver(@config_with_sandbox_mode)
+
+    assert_receive {:fake_sendgrid, %{params: params}}
+    assert params["mail_settings"]["sandbox_mode"]["enable"]
+  end
+
```

I'd like to hear what you think and if a PR should be opened.

**Comment by gullitmiranda** [comment]
@paulcsmith this issue has already been closed with the PR #367, alright?

**Comment by MvanDiemen** [comment]
Ah sorry, yes this is fixed with #367, I'll close this issue.
